### PR TITLE
Simpler emplace

### PR DESCRIPTION
**Pull request by andralex**
The logic in `emplace` copies the logic used by the compiler itself in generating constructors. After sleeping on it for a night I figured I could get most of that logic just work for us by defining a `struct` with only one member and one constructor. Then the rest is taken care of.

### CODE REVIEW

**Comment by adamdruppe** [comment]
Looks like a commit about static assert got mixed in this branch too....

**Comment by andralex** [comment]
Boy I now messet stuff up... hold on...

**Comment by eco** [comment]
Where does the name `kakados` come from? Just curious.

**Comment by andralex** [comment]
I knew it!

That code was in an experimental state. When I introduce experimental names, I use names that would never be plausible in the production codebase, like `crap` or worse. "Caca" means "poop" in Romanian. "Cacados" would be a pretend-Spanish version of it (that doesn't actually exist but sounds Spanish), same as e.g. "los poopos" in English. Now to have additional fun, I also spelled it with "k"s instead of "c"s.

Happened to me in the Facebook codebase... and was seen by my Romanian coworkers :o).

**Comment by andralex** [comment]
Alright I guess I fixed it but will need to wait for https://github.com/D-Programming-Language/phobos/pull/3804.

**Comment by eco** [comment]
"Caca" means poop in English too :)

**Comment by andralex** [comment]
@eco I live and I learn

**Comment by JakobOvrum** [comment]
Looking good! Thanks for these changes, I never understood why `emplace[Ref]` were so messy. I've encountered a fair number of unhandled edge cases with it, but since piling on more spaghetti code isn't scaleable I just shrugged and moved on. Hopefully we can make it more general.

**Comment by andralex** [comment]
BTW this adds documentation examples. But I'm still unhappy about dox - I think all overloads should appear in one place. Will get to that later.

**Comment by andralex** [comment]
tests passed :)

File: std/conv.d [hunk 1]
```diff
@@ -3901,136 +3901,81 @@ Furthermore, emplaceRef optionally takes a type paremeter, which specifies
...
         static assert (is(typeof({static T i;})),
             convFormat("Cannot emplace a %1$s because %1$s.this() is annotated with @disable.", T.stringof));
         emplaceInitializer(chunk);
     }
-
-    // Primitive types, enums, static arrays, dynamic arrays
-    static if (!is(T == struct))
-    void emplaceImpl(Arg)(ref UT chunk, auto ref Arg arg)
+    else static if (
+        !is(T == struct) && Args.length == 1 /* primitives, enums, arrays */
```
Comment (schuetzm) [inline std/conv.d]: I think this needs to check for unions, too.

Comment (andralex) [inline std/conv.d]: Yah, it occurred to me too. That would be an enhancement. I've added a test case for unions (see my next commit) and they seem to work. Could you please post a few tests that fail?

Comment (schuetzm) [inline std/conv.d]: I don't have any. And thinking about it, unions can have neither postblits nor destructors (and can't contain members that do), so they're POD types. They therefore shouldn't need special treatment.

Comment (andralex) [inline std/conv.d]: guess we're ready for the mergearoo!

**Comment by WalterBright** [comment]
How's it doing on:
```
dmd std/conv.d -unittest -cov -main
```
?

**Comment by andralex** [comment]
@WalterBright looks good, only CTFE code and code with `assert(0)` appears uncovered. BTW we shouldn't count paths leading unconditionally to `assert(0)` as uncovered.

I pushed one more commit to nominally improve coverage.

**Comment by andralex** [comment]
Added https://issues.dlang.org/show_bug.cgi?id=15327

**Pull request merged**